rust/src/state_machine/mod.rs

```diff
@@ -0,0 +1,311 @@
+    state_machine::{error::Error, idle::Idle, sum::Sum, sum2::Sum2, update::Update},
+    CoordinatorPublicKey,
+    CoordinatorSecretKey,
+    InitError,
+    PetError,
+    SeedDict,
+    SumDict,
+};
+use std::default::Default;
+use thiserror::Error;
+use tokio::sync::{mpsc, oneshot};
+
+mod error;
+mod idle;
+pub mod requests;
+mod sum;
+mod sum2;
+mod update;
+
+use requests::Request;
+
+#[derive(Debug)]
+pub struct CoordinatorState {
+    pk: CoordinatorPublicKey, // 32 bytes
+    sk: CoordinatorSecretKey, // 32 bytes
+
+    // round parameters
+    sum: f64,
+    update: f64,
+    seed: RoundSeed,
+    min_sum: usize,
+    min_update: usize,
+
+    // round dictionaries
+    /// Dictionary built during the sum phase.
+    sum_dict: SumDict,
+    /// Dictionary built during the update phase.
+    seed_dict: SeedDict,
+    /// Dictionary built during the sum2 phase.
+    mask_dict: MaskDict,
+
+    /// The masking configuration
+    mask_config: MaskConfig,
+
+    /// The aggregated masked model being built in the current round.
+    aggregation: Aggregation,
+}
+
+impl Default for CoordinatorState {
+    fn default() -> Self {
+        let pk = CoordinatorPublicKey::zeroed();
+        let sk = CoordinatorSecretKey::zeroed();
+        let sum = 0.4_f64;
+        let update = 0.5_f64;
+        let seed = RoundSeed::zeroed();
+        let min_sum = 1_usize;
+        let min_update = 3_usize;
+        let sum_dict = SumDict::new();
+        let seed_dict = SeedDict::new();
+        let mask_dict = MaskDict::new();
+        let mask_config = MaskConfig {
+            group_type: GroupType::Prime,
+            data_type: DataType::F32,
+            bound_type: BoundType::B0,
+            model_type: ModelType::M3,
+        };
+        let aggregation = Aggregation::new(mask_config);
+        Self {
+            pk,
+            sk,
+            sum,
+            update,
+            seed,
+            min_sum,
+            min_update,
+            sum_dict,
+            seed_dict,
+            mask_dict,
+            mask_config,
+            aggregation,
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum StateError {
+    #[error("state failed: channel error: {0}")]
+    ChannelError(&'static str),
+    #[error("state failed: round error: {0}")]
+    RoundError(#[from] RoundFailed),
+}
+
+pub struct State<S> {
+    // Inner state
+    _inner: S,
+    // Coordinator state
+    coordinator_state: CoordinatorState,
+    // Request receiver
+    request_rx: mpsc::UnboundedReceiver<Request>,
+}
+
+// Functions that are available to all states
+impl<S> State<S> {
+    /// Receives the next [`Request`].
+    /// Returns [`StateError::ChannelError`] when all senders are dropped.
+    async fn next_request(&mut self) -> Result<Request, StateError> {
+        debug!("received new message");
+        self.request_rx.recv().await.ok_or(StateError::ChannelError(
+            "all message senders have been dropped!",
```

Comment:
I think this is more a shutdown condition than an error condition. Maybe we could have an additional `Shutdown` state instead, if that doesn't require too much changes to the PR?

Reply:
i agree that we should differentiate between recoverable error states and terminal shutdown states.